### DISCUSSION TITLE
Simulate a Smart Meter for a Sungrow Inverter

### DISCUSSION THREAD

**Discussion by Silencer80** [arendst/Tasmota#17207] [Q&A]
Dear Community,

the manufacturers of inverters sell smart meters to determine the flow of electricity from and into the power grid. 

Currently, I am using a Tasmota IR reading head (Hichi IR v1.1) to capture our house electricity meter via SML. All measured values are therefore available. So why should I buy an additional, expensive smart meter and have it installed?

I have the idea to simulate the smart meter. This could be done as follows: 

1. IR reading head reads SML via Tasmota Smart Meter Interface.

2. a NodeMCU with Tasmota is connected to the inverter via RS485 RTU and operates a Modbus bridge.

3. data can be exchanged between the two Tasmota via MQTT. 

Unfortunately I lack knowledge here and would like to kindly ask for support.  

Does the whole thing work at all, or is e.g. the modbus bridge only unidirectional?

Is there a direct way to read the data on Tasmota-1 from SML and provide it on Tasmota-2 as Modbus, or do I have to go via a MQTT broker and transform the data externally if necessary?

I know the Modbus registers the inverter probably wants to query on the simulated smart meter:

The inverter: 
https://ger.sungrowpower.com/productDetail/2209

The inverter supports via RS485 RTU e.g. the following smart meter: Janitza UMG 104.

At Janitza I found the manual with the documentation of all modbus registers, so I probably know what I have to simulate: 

https://www.janitza.de/umg-104-downloads.html?file=files/download/manuals/current/UMG104/janitza-mal-umg104-en.pdf

Do you think this is feasible? 

Many thanks to anyone who can give me helpful tips. 
-Boris

**Comment by barbudor** [comment]
Not natively supported but I would suggest to use an ESP32 based board and to developm the feature ENERGY to SML in Berry

Berry is a python-like scripting language for Tasmota, only available on ESP32
It supports Serial driver and can read the SENSOR/ENERGY messages from Tasmota SML
So you should be able to code that yourself

**Reply by Silencer80** [reply]
Thank you very much as well. Which aspect is not natively supported? Does it refer to the rewriting of the variables? I can certainly do that via my OpenHAB server if it turns out that Berry challenges me too much.

**Reply by barbudor** [reply]
Not natively was refering to move values from reading the SML to sending that out as MODBUS

However Tasmota natively reads your SML and provide the values, that you already achieved
The format however is a SENSOR message

And natively has mqtt to MODBUS bridge

What is missing is the piece in between that you can do in Berry, in Scripting our outside.
Could be on a single Tasmota unit or 2 whatever is more convenient for you

I see that now you have a head start with Gemu's Scripting

**Comment by gemu2015** [comment]
i already started to do exactly this and implemented a client MODBUS mode into scripting.
this would allow to do reading a smart meter and simulate a MODBUS device with one ESP preferably ESP32.
i however did not complete this because the 70 percent rule is no longer required.

i attach my MODBUS client test script to start with.
SML smart meter values are retrieved via global variables in this example.

[modbus_client.txt](https://github.com/arendst/Tasmota/files/10104261/modbus_client.txt)

**Reply by Silencer80** [reply]
Thank you very much! This is a very good starting point. 

I just want to have the data in the iSolarCloud from Sungrow so that the values of grid purchase, delivery and self-consumption are reasonably correct. The 70 percent rule is no longer relevant here in Germany. 

I would like to get this working over the winter, but I am not a good programmer. I know where to plug in the cables and I am good at it. :-)

**Reply by m1k3f15h** [reply]
Hello @Silencer80

Did you got this working already?
I hope I get my Photovoltaic this Spring/Summer..
I wan't tio simulate a Smartmeter für my Heating :)

Mike
